clang/runtime/dpct-rt/include/dpl_extras/memory.h

```diff
@@ -917,34 +928,56 @@ device_pointer<T> malloc_device(const std::size_t num_elements) {
 static inline device_pointer<void> malloc_device(const std::size_t num_bytes) {
   return device_pointer<void>(num_bytes);
 }
-template <typename T>
-device_pointer<T> device_new(device_pointer<T> p, const T &value,
+
+#ifdef DPCT_USM_LEVEL_NONE
+template <typename T, typename U>
+device_pointer<T> device_new(device_pointer<U> p, const T &value,
+                             const std::size_t count) {
+  auto converted_buf = p.buffer.template reinterpret<T>(sycl::range<1>(count));
+  ::std::uninitialized_fill(
+      oneapi::dpl::execution::make_device_policy(dpct::get_default_queue()),
+      oneapi::dpl::begin(converted_buf), oneapi::dpl::end(converted_buf),
+      value);
+  return device_pointer<T>(converted_buf, p.idx);
+}
+// buffer manages lifetime
+template <typename T> void free_device(device_pointer<T> ptr) {}
+#else
+template <typename T, typename U>
+device_pointer<T> device_new(device_pointer<U> p, const T &value,
                              const std::size_t count = 1) {
-  std::vector<T> result(count, value);
-  p.buffer = sycl::buffer<T, 1>(result.begin(), result.end());
-  return p + count;
+  dpct::device_pointer<T> converted_p(static_cast<T *>(p.get()));
+  ::std::uninitialized_fill(
+      oneapi::dpl::execution::make_device_policy(dpct::get_default_queue()),
+      converted_p, converted_p + count, value);
+  return converted_p;
 }
-template <typename T>
-device_pointer<T> device_new(device_pointer<T> p, const std::size_t count = 1) {
+template <typename T> void free_device(device_pointer<T> ptr) {
+  sycl::free(ptr.get(), dpct::get_default_queue());
+}
+#endif
+template <typename T, typename U>
+device_pointer<T> device_new(device_pointer<U> p, const std::size_t count = 1) {
   return device_new(p, T{}, count);
 }
 template <typename T>
 device_pointer<T> device_new(const std::size_t count = 1) {
-  return device_pointer<T>(count);
+  return device_new(device_pointer<T>(sizeof(T) * count), T{}, count);
 }
 
-template <typename T> void free_device(device_pointer<T> ptr) {}
-
 template <typename T>
 typename std::enable_if<!std::is_trivially_destructible<T>::value, void>::type
 device_delete(device_pointer<T> p, const std::size_t count = 1) {
   for (std::size_t i = 0; i < count; ++i) {
```

Comment:
Can we use onedpl::destroy to accelerate the destruction?

Reply:
Yes, I have added this change.